runtime/vm/createramclass.cpp

```diff
@@ -1627,6 +1631,54 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
 	return TRUE;
 }
 
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+/**
+ * Attempts to pre-load classes of fields with Q signatures and no static
+ * access modifier set.
+ *
+ * Caller should not hold the classTableMutex.
+ *
+ * Return TRUE on success. On failure, returns FALSE and sets the
+ * appropriate Java error on the VM.
+ */
+static VMINLINE BOOLEAN
+loadFlattenableFieldValueClasses(J9VMThread *vmThread, J9JavaVM *vm, J9ClassLoader *classLoader, J9ROMClass *romClass, UDATA classPreloadFlags, UDATA packageID, J9Module *module)
+{
+	J9ROMFieldWalkState fieldWalkState = {0};
+
+	/* iterate over fields and load classes of fields marked as QTypes */
+	for (J9ROMFieldShape *field = romFieldsStartDo(romClass, &fieldWalkState); NULL != field; field = romFieldsNextDo(&fieldWalkState)) {
+		const U_32 modifiers = field->modifiers;
+		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(field);
+		if (('Q' == J9UTF8_DATA(signature)[0]) && J9_ARE_NO_BITS_SET(modifiers, J9AccStatic)) {
+			J9Class *valueClass = internalFindClassUTF8(vmThread, &J9UTF8_DATA(signature)[1], J9UTF8_LENGTH(signature)-2, classLoader, classPreloadFlags);
+			if (NULL == valueClass) {
+				return FALSE;
+			} else {
+				J9ROMClass *valueROMClass = valueClass->romClass;
+
+				if (J9_ARE_NO_BITS_SET(valueROMClass->modifiers, J9AccValueType)) {
+					J9UTF8 *badClass = NNSRP_GET(valueROMClass->className, J9UTF8*);
+					setCurrentExceptionNLSWithArgs(vmThread, J9NLS_VM_ERROR_QTYPE_NOT_VALUE_TYPE, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR, J9UTF8_LENGTH(badClass), J9UTF8_DATA(badClass));
+					return FALSE;
+				}
+
+				bool classIsPublic = J9_ARE_ALL_BITS_SET(valueROMClass->modifiers, J9AccPublic);
+
+				if ((!classIsPublic && (packageID != valueClass->packageID))
+					|| (classIsPublic && (J9_VISIBILITY_ALLOWED != checkModuleAccess(vmThread, vm, romClass, module, valueROMClass, valueClass->module, valueClass->packageID, 0)))
+				) {
+					Trc_VM_CreateRAMClassFromROMClass_nestedValueClassNotVisible(vmThread, valueClass, valueClass->classLoader, classLoader);
+					setCurrentExceptionForBadClass(vmThread, J9ROMCLASS_CLASSNAME(valueROMClass), J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSERROR);
+					return FALSE;
+				}
+			}
+		}
+	}
+	return TRUE;
```

Comment:
Please avoid multiple returns - a local and some gotos (if need be) is preferrable.